Frameworks/ImageIO/CGImageSource.mm

```diff
@@ -25,50 +28,327 @@
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Caveat
+ @Notes Current implementation does not support kCGImageSourceShouldAllowFloat & kCGImageSourceShouldCache 
+        when passed in as options dictionary keys
 */
 CGImageRef CGImageSourceCreateImageAtIndex(CGImageSourceRef isrc, size_t index, CFDictionaryRef options) {
-    UNIMPLEMENTED();
-    return StubReturn();
+    RETURN_NULL_IF(!isrc);
+    NSData* imageData = ((ImageSource*)isrc).data;
+    RETURN_NULL_IF(!imageData);
+    RETURN_NULL_IF(index > (CGImageSourceGetCount(isrc) - 1));
+
+    MULTI_QI imageQueryInterface = {0};
+    static const GUID IID_IWICImagingFactory = {0xec5ec8a9,0xc395,0x4314,0x9c,0x77,0x54,0xd7,0xa9,0x35,0xff,0x70};
+    imageQueryInterface.pIID = &IID_IWICImagingFactory;
+    RETURN_NULL_IF_FAILED(
+        CoCreateInstanceFromApp(CLSID_WICImagingFactory, nullptr, CLSCTX_INPROC_SERVER, nullptr, 1, &imageQueryInterface));
+
+    ComPtr<IWICImagingFactory> imageFactory = (IWICImagingFactory*)imageQueryInterface.pItf;
+    ComPtr<IWICStream> imageStream;
+    RETURN_NULL_IF_FAILED(imageFactory->CreateStream(&imageStream));
+
+    unsigned char* imageByteArray = (unsigned char*)[imageData bytes];
+    int imageLength = [imageData length];
+    RETURN_NULL_IF_FAILED(imageStream->InitializeFromMemory(imageByteArray, imageLength));
+
+    if (options && CFDictionaryContainsKey(options, kCGImageSourceShouldCache)) {
+        UNIMPLEMENTED_WITH_MSG("kCGImageSourceShouldCache is not supported in current implementation.");
+    }
+
+    ComPtr<IWICBitmapDecoder> imageDecoder;                
+    RETURN_NULL_IF_FAILED(imageFactory->CreateDecoderFromStream(imageStream.Get(), nullptr, WICDecodeMetadataCacheOnDemand, &imageDecoder));
+
+    ComPtr<IWICBitmapFrameDecode> imageFrame;
+    RETURN_NULL_IF_FAILED(imageDecoder->GetFrame(index, &imageFrame));
+
+    unsigned int frameWidth = 0;
+    unsigned int frameHeight = 0;
+    RETURN_NULL_IF_FAILED(imageFrame->GetSize(&frameWidth, &frameHeight));
+
+    ComPtr<IWICFormatConverter> imageFormatConverter;
+    RETURN_NULL_IF_FAILED(imageFactory->CreateFormatConverter(&imageFormatConverter));
+
+    if (options && CFDictionaryContainsKey(options, kCGImageSourceShouldAllowFloat)) {
+        UNIMPLEMENTED_WITH_MSG("kCGImageSourceShouldAllowFloat is not supported in current implementation.");
+    }
+
+    RETURN_NULL_IF_FAILED(imageFormatConverter->Initialize(imageFrame.Get(), 
+                                                           GUID_WICPixelFormat32bppRGBA,
+                                                           WICBitmapDitherTypeNone, 
+                                                           nullptr, 
+                                                           0.f, 
+                                                           WICBitmapPaletteTypeCustom));
+
+    const unsigned int frameSize = frameWidth * frameHeight * 4;
+    std::vector<unsigned char> frameByteVector(frameSize);
```

Comment:
std::vector doesn't let you transfer ownership of its buffer like this.  I think you'll need std::unique_ptr<unsigned char[]>, then call .release() on it after the successful copy into frameData.